// fontstyles/index.js
let packageJson;

try {
  packageJson = require('../../package.json');
} catch (error) {
  packageJson = require('./package.json');
}

function apply(text, fontMap) {
  if (!text || text.length === 0) {
    return "";
  }

  return text.replace(/[a-zA-Z0-9]/g, (char) => {
    const fonty = fontMap[char];
    return fonty ? fonty : char;
  });
}
    
const thin = {
  a: 'ð–º', b: 'ð–»', c: 'ð–¼', d: 'ð–½', e: 'ð–¾', f: 'ð–¿',
  g: 'ð—€', h: 'ð—', i: 'ð—‚', j: 'ð—ƒ', k: 'ð—„', l: 'ð—…',
  m: 'ð—†', n: 'ð—‡', o: 'ð—ˆ', p: 'ð—‰', q: 'ð—Š', r: 'ð—‹',
  s: 'ð—Œ', t: 'ð—', u: 'ð—Ž', v: 'ð—', w: 'ð—', x: 'ð—‘',
  y: 'ð—’', z: 'ð—“',
  A: 'ð– ', B: 'ð–¡', C: 'ð–¢', D: 'ð–£', E: 'ð–¤', F: 'ð–¥',
  G: 'ð–¦', H: 'ð–§', I: 'ð–¨', J: 'ð–©', K: 'ð–ª', L: 'ð–«',
  M: 'ð–¬', N: 'ð–­', O: 'ð–®', P: 'ð–¯', Q: 'ð–°', R: 'ð–±',
  S: 'ð–²', T: 'ð–³', U: 'ð–´', V: 'ð–µ', W: 'ð–¶', X: 'ð–·',
  Y: 'ð–¸', Z: 'ð–¹',
  0: 'ðŸ¢', 1: 'ðŸ£', 2: 'ðŸ¤', 3: 'ðŸ¥', 4: 'ðŸ¦',
  5: 'ðŸ§', 6: 'ðŸ¨', 7: 'ðŸ©', 8: 'ðŸª', 9: 'ðŸ«'
};

const roman = {
  a: 'ð‘Ž', b: 'ð‘', c: 'ð‘', d: 'ð‘‘', e: 'ð‘’', f: 'ð‘“',
  g: 'ð‘”', h: 'â„Ž', i: 'ð‘–', j: 'ð‘—', k: 'ð‘˜', l: 'ð‘™',
  m: 'ð‘š', n: 'ð‘›', o: 'ð‘œ', p: 'ð‘', q: 'ð‘ž', r: 'ð‘Ÿ',
  s: 'ð‘ ', t: 'ð‘¡', u: 'ð‘¢', v: 'ð‘£', w: 'ð‘¤', x: 'ð‘¥',
  y: 'ð‘¦', z: 'ð‘§',
  A: 'ð´', B: 'ðµ', C: 'ð¶', D: 'ð·', E: 'ð¸', F: 'ð¹',
  G: 'ðº', H: 'ð»', I: 'ð¼', J: 'ð½', K: 'ð¾', L: 'ð¿',
  M: 'ð‘€', N: 'ð‘', O: 'ð‘‚', P: 'ð‘ƒ', Q: 'ð‘„', R: 'ð‘…',
  S: 'ð‘†', T: 'ð‘‡', U: 'ð‘ˆ', V: 'ð‘‰', W: 'ð‘Š', X: 'ð‘‹',
  Y: 'ð‘Œ', Z: 'ð‘',
  0: 'ðŸŽ', 1: 'ðŸ', 2: 'ðŸ', 3: 'ðŸ‘', 4: 'ðŸ’',
  5: 'ðŸ“', 6: 'ðŸ”', 7: 'ðŸ•', 8: 'ðŸ–', 9: 'ðŸ—'
};

const strike = {
  a: 'Ì¶aÌ¶', b: 'Ì¶bÌ¶', c: 'Ì¶cÌ¶', d: 'Ì¶dÌ¶', e: 'Ì¶eÌ¶', f: 'Ì¶fÌ¶',
  g: 'Ì¶gÌ¶', h: 'Ì¶hÌ¶', i: 'Ì¶iÌ¶', j: 'Ì¶jÌ¶', k: 'Ì¶kÌ¶', l: 'Ì¶lÌ¶',
  m: 'Ì¶mÌ¶', n: 'Ì¶nÌ¶', o: 'Ì¶oÌ¶', p: 'Ì¶pÌ¶', q: 'Ì¶qÌ¶', r: 'Ì¶rÌ¶',
  s: 'Ì¶sÌ¶', t: 'Ì¶tÌ¶', u: 'Ì¶uÌ¶', v: 'Ì¶vÌ¶', w: 'Ì¶wÌ¶', x: 'Ì¶xÌ¶',
  y: 'Ì¶yÌ¶', z: 'Ì¶zÌ¶',
  A: 'Ì¶AÌ¶', B: 'Ì¶BÌ¶', C: 'Ì¶CÌ¶', D: 'Ì¶DÌ¶', E: 'Ì¶EÌ¶', F: 'Ì¶FÌ¶',
  G: 'Ì¶GÌ¶', H: 'Ì¶HÌ¶', I: 'Ì¶IÌ¶', J: 'Ì¶JÌ¶', K: 'Ì¶KÌ¶', L: 'Ì¶LÌ¶',
  M: 'Ì¶MÌ¶', N: 'Ì¶NÌ¶', O: 'Ì¶OÌ¶', P: 'Ì¶PÌ¶', Q: 'Ì¶QÌ¶', R: 'Ì¶RÌ¶',
  S: 'Ì¶SÌ¶', T: 'Ì¶TÌ¶', U: 'Ì¶UÌ¶', V: 'Ì¶VÌ¶', W: 'Ì¶WÌ¶', X: 'Ì¶XÌ¶',
  Y: 'Ì¶YÌ¶', Z: 'Ì¶ZÌ¶',
  0: 'Ì¶0Ì¶', 1: 'Ì¶1Ì¶', 2: 'Ì¶2Ì¶', 3: 'Ì¶3Ì¶', 4: 'Ì¶4Ì¶',
  5: 'Ì¶5Ì¶', 6: 'Ì¶6Ì¶', 7: 'Ì¶7Ì¶', 8: 'Ì¶8Ì¶', 9: 'Ì¶9Ì¶'
};
  
const bold = {
  a: 'ð—®', b: 'ð—¯', c: 'ð—°', d: 'ð—±', e: 'ð—²', f: 'ð—³',
  g: 'ð—´', h: 'ð—µ', i: 'ð—¶', j: 'ð—·', k: 'ð—¸', l: 'ð—¹',
  m: 'ð—º', n: 'ð—»', o: 'ð—¼', p: 'ð—½', q: 'ð—¾', r: 'ð—¿',
  s: 'ð˜€', t: 'ð˜', u: 'ð˜‚', v: 'ð˜ƒ', w: 'ð˜„', x: 'ð˜…',
  y: 'ð˜†', z: 'ð˜‡',
  A: 'ð—”', B: 'ð—•', C: 'ð—–', D: 'ð——', E: 'ð—˜', F: 'ð—™',
  G: 'ð—š', H: 'ð—›', I: 'ð—œ', J: 'ð—', K: 'ð—ž', L: 'ð—Ÿ',
  M: 'ð— ', N: 'ð—¡', O: 'ð—¢', P: 'ð—£', Q: 'ð—¤', R: 'ð—¥',
  S: 'ð—¦', T: 'ð—§', U: 'ð—¨', V: 'ð—©', W: 'ð—ª', X: 'ð—«',
  Y: 'ð—¬', Z: 'ð—­',
  0: 'ðŸ¬', 1: 'ðŸ­', 2: 'ðŸ®', 3: 'ðŸ¯', 4: 'ðŸ°',
  5: 'ðŸ±', 6: 'ðŸ²', 7: 'ðŸ³', 8: 'ðŸ´', 9: 'ðŸµ'
};

const underline = {
  a: 'aÌ²', b: 'bÌ²', c: 'cÌ²', d: 'dÌ²', e: 'eÌ²', f: 'fÌ²',
g: 'gÌ²', h: 'hÌ²', i: 'iÌ²', j: 'jÌ²', k: 'kÌ²', l: 'lÌ²',
m: 'mÌ²', n: 'nÌ²', o: 'oÌ²', p: 'pÌ²', q: 'qÌ²', r: 'rÌ²',
s: 'sÌ²', t: 'tÌ²', u: 'uÌ²', v: 'vÌ²', w: 'wÌ²', x: 'xÌ²',
y: 'yÌ²', z: 'zÌ²',
A: 'AÌ²', B: 'BÌ²', C: 'CÌ²', D: 'DÌ²', E: 'EÌ²', F: 'FÌ²',
G: 'GÌ²', H: 'HÌ²', I: 'IÌ²', J: 'JÌ²', K: 'KÌ²', L: 'LÌ²',
M: 'MÌ²', N: 'NÌ²', O: 'OÌ²', P: 'PÌ²', Q: 'QÌ²', R: 'RÌ²',
S: 'SÌ²', T: 'TÌ²', U: 'UÌ²', V: 'VÌ²', W: 'WÌ²', X: 'XÌ²',
Y: 'YÌ²', Z: 'ZÌ²',
0: '0Ì²', 1: '1Ì²', 2: '2Ì²', 3: '3Ì²', 4: '4Ì²',
5: '5Ì²', 6: '6Ì²', 7: '7Ì²', 8: '8Ì²', 9: '9Ì²'
};

const monospace = {
  'a': 'ðšŠ', 'b': 'ðš‹', 'c': 'ðšŒ', 'd': 'ðš', 'e': 'ðšŽ', 'f': 'ðš',
  'g': 'ðš', 'h': 'ðš‘', 'i': 'ðš’', 'j': 'ðš“', 'k': 'ðš”', 'l': 'ðš•',
  'm': 'ðš–', 'n': 'ðš—', 'o': 'ðš˜', 'p': 'ðš™', 'q': 'ðšš', 'r': 'ðš›',
  's': 'ðšœ', 't': 'ðš', 'u': 'ðšž', 'v': 'ðšŸ', 'w': 'ðš ', 'x': 'ðš¡',
  'y': 'ðš¢', 'z': 'ðš£',
  'A': 'ð™°', 'B': 'ð™±', 'C': 'ð™²', 'D': 'ð™³', 'E': 'ð™´', 'F': 'ð™µ',
  'G': 'ð™¶', 'H': 'ð™·', 'I': 'ð™¸', 'J': 'ð™¹', 'K': 'ð™º', 'L': 'ð™»',
  'M': 'ð™¼', 'N': 'ð™½', 'O': 'ð™¾', 'P': 'ð™¿', 'Q': 'ðš€', 'R': 'ðš',
  'S': 'ðš‚', 'T': 'ðšƒ', 'U': 'ðš„', 'V': 'ðš…', 'W': 'ðš†', 'X': 'ðš‡',
  'Y': 'ðšˆ', 'Z': 'ðš‰',
  '0': 'ðŸ¶', '1': 'ðŸ·', '2': 'ðŸ¸', '3': 'ðŸ¹', '4': 'ðŸº',
  '5': 'ðŸ»', '6': 'ðŸ¼', '7': 'ðŸ½', '8': 'ðŸ¾', '9': 'ðŸ¿'
};

const squarebox = {
  'a': 'ðŸ„°', 'b': 'ðŸ„±', 'c': 'ðŸ„²', 'd': 'ðŸ„³', 'e': 'ðŸ„´', 'f': 'ðŸ„µ',
  'g': 'ðŸ„¶', 'h': 'ðŸ„·', 'i': 'ðŸ„¸', 'j': 'ðŸ„¹', 'k': 'ðŸ„º', 'l': 'ðŸ„»',
  'm': 'ðŸ„¼', 'n': 'ðŸ„½', 'o': 'ðŸ„¾', 'p': 'ðŸ„¿', 'q': 'ðŸ…€', 'r': 'ðŸ…',
  's': 'ðŸ…‚', 't': 'ðŸ…ƒ', 'u': 'ðŸ…„', 'v': 'ðŸ……', 'w': 'ðŸ…†', 'x': 'ðŸ…‡',
  'y': 'ðŸ…ˆ', 'z': 'ðŸ…‰',
  'A': 'ðŸ„°', 'B': 'ðŸ„±', 'C': 'ðŸ„²', 'D': 'ðŸ„³', 'E': 'ðŸ„´', 'F': 'ðŸ„µ',
  'G': 'ðŸ„¶', 'H': 'ðŸ„·', 'I': 'ðŸ„¸', 'J': 'ðŸ„¹', 'K': 'ðŸ„º', 'L': 'ðŸ„»',
  'M': 'ðŸ„¼', 'N': 'ðŸ„½', 'O': 'ðŸ„¾', 'P': 'ðŸ„¿', 'Q': 'ðŸ…€', 'R': 'ðŸ…',
  'S': 'ðŸ…‚', 'T': 'ðŸ…ƒ', 'U': 'ðŸ…„', 'V': 'ðŸ……', 'W': 'ðŸ…†', 'X': 'ðŸ…‡',
  'Y': 'ðŸ…ˆ', 'Z': 'ðŸ…‰',
  '0': '0âƒ£', '1': '1âƒ£', '2': '2âƒ£', '3': '3âƒ£', '4': '4âƒ£',
  '5': '5âƒ£', '6': '6âƒ£', '7': '7âƒ£', '8': '8âƒ£', '9': '9âƒ£'
};

// Not yet added to the Origin ðŸ‘‡
const bubble = {
  'a': 'â“', 'b': 'â“‘', 'c': 'â“’', 'd': 'â““', 'e': 'â“”', 'f': 'â“•',
  'g': 'â“–', 'h': 'â“—', 'i': 'â“˜', 'j': 'â“™', 'k': 'â“š', 'l': 'â“›',
  'm': 'â“œ', 'n': 'â“', 'o': 'â“ž', 'p': 'â“Ÿ', 'q': 'â“ ', 'r': 'â“¡',
  's': 'â“¢', 't': 'â“£', 'u': 'â“¤', 'v': 'â“¥', 'w': 'â“¦', 'x': 'â“§',
  'y': 'â“¨', 'z': 'â“©',
  'A': 'â’¶', 'B': 'â’·', 'C': 'â’¸', 'D': 'â’¹', 'E': 'â’º', 'F': 'â’»',
  'G': 'â’¼', 'H': 'â’½', 'I': 'â’¾', 'J': 'â’¿', 'K': 'â“€', 'L': 'â“',
  'M': 'â“‚', 'N': 'â“ƒ', 'O': 'â“„', 'P': 'â“…', 'Q': 'â“†', 'R': 'â“‡',
  'S': 'â“ˆ', 'T': 'â“‰', 'U': 'â“Š', 'V': 'â“‹', 'W': 'â“Œ', 'X': 'â“',
  'Y': 'â“Ž', 'Z': 'â“',
  '0': 'â“ª', '1': 'â‘ ', '2': 'â‘¡', '3': 'â‘¢', '4': 'â‘£',
  '5': 'â‘¤', '6': 'â‘¥', '7': 'â‘¦', '8': 'â‘§', '9': 'â‘¨'
};

const boldItalic = {
  ð™–: 'ð™–', ð™—: 'ð™—', ð™˜: 'ð™˜', ð™™: 'ð™™', ð™š: 'ð™š', ð™›: 'ð™›',
  ð™œ: 'ð™œ', ð™: 'ð™', ð™ž: 'ð™ž', ð™Ÿ: 'ð™Ÿ', ð™ : 'ð™ ', ð™¡: 'ð™¡',
  ð™¢: 'ð™¢', ð™£: 'ð™£', ð™¤: 'ð™¤', ð™¥: 'ð™¥', ð™¦: 'ð™¦', ð™§: 'ð™§',
  ð™¨: 'ð™¨', ð™©: 'ð™©', ð™ª: 'ð™ª', ð™«: 'ð™«', ð™¬: 'ð™¬', ð™­: 'ð™­',
  ð™®: 'ð™®', ð™¯: 'ð™¯',
  ð˜¼: 'ð˜¼', ð˜½: 'ð˜½', ð˜¾: 'ð˜¾', ð˜¿: 'ð˜¿', ð™€: 'ð™€', ð™: 'ð™',
  ð™‚: 'ð™‚', ð™ƒ: 'ð™ƒ', ð™„: 'ð™„', ð™…: 'ð™…', ð™†: 'ð™†', ð™‡: 'ð™‡',
  ð™ˆ: 'ð™ˆ', ð™‰: 'ð™‰', ð™Š: 'ð™Š', ð™‹: 'ð™‹', ð™Œ: 'ð™Œ', ð™: 'ð™',
  ð™Ž: 'ð™Ž', ð™: 'ð™', ð™: 'ð™', ð™‘: 'ð™‘', ð™’: 'ð™’', ð™“: 'ð™“',
  ð™”: 'ð™”', ð™•: 'ð™•',
  0: 'ðŸ¬', 1: 'ðŸ­', 2: 'ðŸ®', 3: 'ðŸ¯', 4: 'ðŸ°',
  5: 'ðŸ±', 6: 'ðŸ²', 7: 'ðŸ³', 8: 'ðŸ´', 9: 'ðŸµ',
};

const italic = {
  a: 'ð˜¢', b: 'ð˜£', c: 'ð˜¤', d: 'ð˜¥', e: 'ð˜¦', f: 'ð˜§',
  g: 'ð˜¨', h: 'ð˜©', i: 'ð˜ª', j: 'ð˜«', k: 'ð˜¬', l: 'ð˜­',
  m: 'ð˜®', n: 'ð˜¯', o: 'ð˜°', p: 'ð˜±', q: 'ð˜²', r: 'ð˜³',
  s: 'ð˜´', t: 'ð˜µ', u: 'ð˜¶', v: 'ð˜·', w: 'ð˜¸', x: 'ð˜¹',
  y: 'ð˜º', z: 'ð˜»',
  A: 'ð˜ˆ', B: 'ð˜‰', C: 'ð˜Š', D: 'ð˜‹', E: 'ð˜Œ', F: 'ð˜',
  G: 'ð˜Ž', H: 'ð˜', I: 'ð˜', J: 'ð˜‘', K: 'ð˜’', L: 'ð˜“',
  M: 'ð˜”', N: 'ð˜•', O: 'ð˜–', P: 'ð˜—', Q: 'ð˜˜', R: 'ð˜™',
  S: 'ð˜š', T: 'ð˜›', U: 'ð˜œ', V: 'ð˜', W: 'ð˜ž', X: 'ð˜Ÿ',
  Y: 'ð˜ ', Z: 'ð˜¡',
  0: 'ðŸ¬', 1: 'ðŸ­', 2: 'ðŸ®', 3: 'ðŸ¯', 4: 'ðŸ°',
  5: 'ðŸ±', 6: 'ðŸ²', 7: 'ðŸ³', 8: 'ðŸ´', 9: 'ðŸµ'
};

const { revert } = require('./mods/origin.js');

module.exports = {
  thin: (text) => apply(text, thin),
  bold: (text) => apply(text, bold),
  origin: (text) => revert(text),
  italic: (text) => apply(text, italic),
  strike: (text) => apply(text, strike),
  underline: (text) => apply(text, underline),
  monospace: (text) => apply(text, monospace),
  boldItalic: (text) => apply(text, boldItalic),
  roman: (text) => apply(text, roman),
  bubble: (text) => apply(text, bubble),
  squarebox: (text) => apply(text, squarebox)
};
